### PR TITLE
Profile hardening

### PR DESCRIPTION
**Pull request by qujia7**
example of profile hardening with tsz

### CODE REVIEW

File: solenspipe/solenspipe.py [hunk 1]
```diff
@@ -121,14 +121,14 @@ def get_tempura_norms(est1,est2,ucls,tcls,lmin,lmax,mlmax):
...
-        if bh:
+        if bh and est2=='SRC':
             Nl_g_bh = bias_hardened_n0(Nl_phi_g,Als['src'],R_src_tt) * (ls*(ls+1.)/2.)**2.
         else:
             Nl_g_bh = None
     return bh,ls,Als,R_src_tt,Nl_g,Nl_c,Nl_g_bh
 
 
-def get_qfunc(px,ucls,mlmax,est1,Al1=None,est2=None,Al2=None,R12=None):
+def get_qfunc(px,ucls,mlmax,est1,Al1=None,est2=None,Al2=None,R12=None,profile=None):
```
Comment (NiallMac) [inline solenspipe/solenspipe.py:131]: Looks good - just need to add the profile argument to the Parameters list in the docstring

**Comment by NiallMac** [comment]
Nice stuff Frank. One comment - I'm a bit unsure about the changes to `solenspipe.get_tempura_norms` (actually it's a bit unclear to me what we want from this function in general). I think this was meant to be a convenience function that just calculates all the relevant normalizations/responses - it looks like that was the original intention (@msyriac?) since the current behaviour is that it always calculates the bias-hardening response when doing TT or MV. With your change it only does this when est2='SRC' which somewhat changes the usage e.g. I think this would break the current verify.py? Would it make sense to add a `profile` argument to this function such that the current default behaviour remains the same, and if profile is provided it additionally calculates the normalization and response for the provided profile (or even a list of provided profiles)?

File: examples/checkPH.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    #normalization for TT
+    Als = pytempura.get_norms(est_norm_list,ucls,tcls,lmin,lmax,k_ellmax=mlmax)
+
+    #tcls_norm["TT"][0:3000] /= (profile**2)[0:3000]
+    tcls["TT"][0:3000] /= profile[0:3000]
+    
+
+with bench.show("NORM"):
+    # Get response for TSZ profile
+    R_src_tt = pytempura.get_cross('SRC','TT',ucls,tcls,lmin,lmax,k_ellmax=mlmax)/=profile[0:3000]
```
Comment (NiallMac) [inline examples/checkPH.py]: I guess this should be a "/" rather than "/="?

**Comment by NiallMac** [comment]
I guess my other question about this stuff is whether going forward, we want to implement the profile-hardening response/normalization stuff by adjusting the tcls passed to pytempura and applying correction prefactors, as Frank has cleverly done here, or should we update pytempura directly to accept a profile? Maybe the latter seems a bit neater going forward? @toshiyan?

**Comment by NiallMac** [comment]
(@qujia7 @msyriac happy to push an updated `get_tempura_norms` to this branch based on my comment above if you think that sounds sensible)

File: examples/checkPH.py (hunk 2 above)
Comment (NiallMac) [reply examples/checkPH.py]: (pushed a fix assuming that is the case)

File: examples/checkPH.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+    # Get CMB Cls for response and total Cls corresponding to a noiseless configuration
+    ucls,tcls = utils.get_theory_dicts(grad=grad)
+    est_norm_list = ['TT','src']
+
+    #normalization for TT
+    Als = pytempura.get_norms(est_norm_list,ucls,tcls,lmin,lmax,k_ellmax=mlmax)
+
+    #tcls_norm["TT"][0:3000] /= (profile**2)[0:3000]
+    tcls["TT"][0:3000] /= profile[0:3000]
```
Comment (NiallMac) [inline examples/checkPH.py:71]: Should all these `[:3000]` ranges be `[:args.lmax+1]`?

**Comment by msyriac** [comment]
> (@qujia7 @msyriac happy to push an updated `get_tempura_norms` to this branch based on my comment above if you think that sounds sensible)

Updating pytempura would be ideal. Feel free to PR!

**Pull request closed without merging**